### ISSUE TITLE
Spurious error generated when inner function contains `P.args` annotation and decorator

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5913]
Paste the following code into the editor. Then select the `u` in the expression `fn(u, *args, **kwargs)` and type `u`. A spurious error will be reported against the `args` subexpression. This is because the `args` declaration requires the evaluation of `P.args`, and this requires the type evaluator to determine the scope of `P`. This causes it to evaluate the type of `wrapper` to determine if `P` is a type parameter for this function. This applies the decorator `wraps`, which requires the evaluation of the return type. Since the return type isn't declared, it must be inferred from the return statement. This requires evaluation of `args`, which is already being evaluated. A circular dependency is therefore reported.

```python
from functools import wraps
from typing import Any, Callable, TypeVar
from pydantic import BaseModel
from typing_extensions import Concatenate, ParamSpec


class User(BaseModel):
    id: int


StringKeyDict = dict[str, Any]
P = ParamSpec("P")
R = TypeVar("R")

ModelFunc = Callable[Concatenate[User, P], R]
DictFunc = Callable[Concatenate[StringKeyDict, P], R]


def accept_dict(fn: ModelFunc[P, R]) -> DictFunc[P, R]:
    @wraps(fn)
    def wrapper(x: StringKeyDict, *args: P.args, **kwargs: P.kwargs):
        u = User(**x)
        return fn(u, *args, **kwargs)

    return wrapper


@accept_dict
def expects_pydantic(x: User) -> int:
    return x.id


expects_pydantic({"id": 1})
```

**Comment by samiur** [comment]
Thanks for the fix @erictraut! Ran into this yesterday and for a while was like "what am I doing wrong???". Checked with mypy and saw that it's actually fine, so realized it must have been a pyright bug.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.327, which I just published. It will also be included in an upcoming release of pylance.
